Add module for computing river flood footprints from GloFAS river discharge data

peanutfun opened the pull request:
The module includes a data pipeline which automatically downloads GloFAS river discharge data and transforms it to flood footprints, which in turn can be transfered to``Hazard`` or ``RiverFlood`` objects.

The module introduces a few new dependencies. They have to be installed into the Jenkins test environment for the builds to succeed:
- [dantro](https://gitlab.com/utopia-project/dantro) for the data pipeline
- [cdsapi](https://github.com/ecmwf/cdsapi) for downloading data from the Copernicus Data Store
- [ruamel.yaml](https://pypi.org/project/ruamel.yaml/) for reading YAML files. This one is already a dependency of dantro.

Major changes in CLIMADA Petals:
* Add the `climada_petals.hazard.rf_glofas` subpackage
* Add `climada_petals.util.cds_glofas_downloader` utility functions
* Add module documentation

To do:
- [x] Tutorial
- [x] FLOPROS database integration
- [x] Some more tests
- [x] Fix linter issues

peanutfun commented:
@emanuel-schmid Could you have a look at why the new dependencies are not found in the checks?

aleeciu commented:
As briefly discussed with @peanutfun , I will start with using the module and report on the experience. Then we clarify whether I should also look into the code or not.

peanutfun commented:
@emanuel-schmid For the tests I need the package `xesmf`, which is not installable via pip. Could you add it to the petals testing environment?

emanuel-schmid commented:
sure. 😃 see 2590349

peanutfun commented on script/jenkins/branches/Jenkinsfile:
```diff
@@ -23,6 +23,8 @@ pipeline {
             export PATH=$PATH:$CONDAPATH
             source activate petals_env
+            mamba install xesmf -c conda-forge -y
+            pip install cdsapi dantro matplotlib>3.4 rioxarray ruamel.yaml
```
Try upgrading to ensure matplotlib>3.4 is installed
```suggestion
            pip install --upgrade cdsapi dantro matplotlib>3.4 rioxarray ruamel.yaml
```

peanutfun commented on script/jenkins/branches/Jenkinsfile:
```diff
@@ -23,6 +23,8 @@ pipeline {
             export PATH=$PATH:$CONDAPATH
             source activate petals_env
+            mamba install xesmf -c conda-forge -y
+            pip install --upgrade cdsapi dantro matplotlib>3.4 rioxarray ruamel.yaml
```
```suggestion
            pip install --upgrade cdsapi dantro matplotlib==3.5.* rioxarray ruamel.yaml
```

tovogt commented on climada_petals/util/cds_glofas_downloader.py:
```diff
@@ -0,0 +1,256 @@
+        "leadtime_hour": list(map(str, range(1, 31))),
+    },
+}
+"""Default request keyword arguments to be updated by the user requests"""
```
I don't understand why you would have a `util` module for this functionality in the first place. The functions in `cds_glofas_downloader` are only used in the `climada_petals.hazard.rf_glofas` subpackage, so why not create it as a new module in that subpackage? The whole module seems very application-specific to me, it even has `glofas` in its name. I think it's completely safe to move it to `climada_petals.hazard.rf_glofas.cds_downloader`.

peanutfun replied:
True. I envisaged that this module can be easily extended to download other data than those of GloFAS from the CDS, hence I put it there. I'll move it.

peanutfun replied:
I moved the download module into the hazard submodule.

tovogt reviewed (commented):
Great work, @peanutfun !

It was also very interesting for me personally to learn about the dantro framework.

However, I'm wondering whether the workflow with the customized yml files is overly complicated for most users. The typical application seems to be that a user is interested in a particular region and a particular time period, they might be working with historical or forecast data, and they might want to be able to control whether Flopros data is applied or not. That's a very compact amount of configuration parameters and I can easily imagine a function that just takes these parameters and then creates the correct yml file, executes the dantro transformations and returns the list of RiverFlood hazard objects - without the user even knowing that dantro is used internally.

It's really invaluable that you documented in much detail how dantro allows users to create their custom data pipelines etc. But I would rather have this as a second chapter, for power users. In the tutorial notebook, and in the docs, I don't understand why anyone should be bothered with this in the first few paragraphs.

As mentioned in the code, I also think that a lot of the data that is downloaded from permanent locations and then stored in or extracted to the CLIMADA data directory, should be downloaded (and extracted) automatically by your code if needed instead of having the user create a directory, visit the web sites and click on download buttons. For an example how this is done, see the `_get_dist_to_coast_nasa_tif` function in the `climada.util.coordinates` module.

peanutfun commented on doc/glofas_rf.rst:
```diff
@@ -0,0 +1,449 @@
+
+  Now the file is created and can be opened with your favorite text editor.
+
+Create Data Directory
+^^^^^^^^^^^^^^^^^^^^^
+
+1. Enter the CLIMADA data directory.
+   If you did not change the default configuration, this is located in ``$HOME/climada/data``.
+
+  If this directory does not exist, make sure you activated the CLIMADA environment and execute
+
+  .. code-block:: shell
+
+    python -c "import climada"
+
+2. Create a new directory called ``glofas-computation`` in the data directory.
+
+Download Prepared Datasets from the ETH Research Collection
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The data required for flood footprint computation has been uploaded to the `ETH Research Collection <https://www.research-collection.ethz.ch/>`_ for your convenience:
+
+* `Gumbel distribution fit parameters for historical GloFAS river discharge data <https://doi.org/10.3929/ethz-b-000587207>`_
+* `Flood hazard map of the World <https://doi.org/10.3929/ethz-b-000587202>`_
+
+Place the downloaded files in the ``glofas-computation`` data directory you created in the last section.
+If you did that, you can directly proceed with :ref:`computing flood footprints <compute>`.
+
+If not, you will have to follow the next steps to compute the input data yourself:
```
Yes, I will add helper functions for that.

peanutfun replied:
Helper functions triggered by `setup_all()` now download all data automatically.

peanutfun commented on doc/glofas_rf.rst:
```diff
@@ -0,0 +1,449 @@
+If you cannot download the data, you can re-create it on your machine using the :py:func:`climada_petals.hazard.rf_glofas.setup` function.
+To do so, follow the optional instructions below.
+
+First, download the global flood hazard map source data.
+Freely accessible global flood hazard maps are available from the Joint Research Centre (JRC) Data Catalogue as GeoTIFF files: https://data.jrc.ec.europa.eu/collection/id-0054
+
+1. Download all "Flood hazard map of the World" from the JRC Data Catalogue.
+   Click on each respective entry, navigate to the "Data access" section and click "Download".
+2. Unzip the downloaded files.
+3. Copy or move the unzipped folders into the data directory.
```
Correct. I think this part is now obsolete because I uploaded the prepared files to the ETH research collection and will provide helper functions that automatically download them (see my other comment)

tovogt replied:
But you might want to have reproducible code to produce the pre-computed files in the ETH research collection. So, I would still suggest to implement this.

peanutfun replied:
These files are downloaded and merged now with `setup_all`

peanutfun commented:
@tovogt Thanks a lot for taking your time and reviewing the module!

> However, I'm wondering whether the workflow with the customized yml files is overly complicated for most users.

I think you are right. But I was and still am hesitant to wrap all this functionality into one or two single function calls. The issue is that there are way to many parameters to tweak. You suggest breaking down the functionality and I can see that this might work. However, I fear that users might quickly create a pipeline that requires, e.g., rechunking of the data or even writing intermediate files to disk to free up memory for remaining somewhat efficient. On the other hand, I lately improved the performance of some critical operations and I haven't tested yet if the aforementioned tweaks are still required to achieve reasonable performance.

So, how do we go about that? I presume that writing wrapper functions needs quite a bit more exploring and testing, and I currently don't have time for that. So we either
1. keep the PR as is until I/we get around to doing that, or
2. we merge with the current feature set and create a follow-up issue/PR for writing suitable wrapper functions.

What do you think?

peanutfun commented on doc/glofas_rf.rst:
```diff
@@ -0,0 +1,449 @@
+For computing these return periods, we require an extreme value distribution at every point.
+In practice, we fit such distributions from the historical discharge data.
+
+Note that computing river flood hazards with this module still is computationally expensive.
```
Good question. The "slowest" part so far is the last step, the interpolation of the flood depth. This is due to calling `np.vectorize`, which essentially just loops over the function. I have a local version where I use explicit vectorization and the whole thing now runs much faster. I can currently give no exact timings, though.

tovogt replied:
There's no need for "exact timings", just give an idea of what to expect. I mean, if you can't specify this a bit more, the sentence is really meaningless, right?

The sentence should say something like: "computing river flood hazards with this module is still computationally expensive compared to other CLIMADA functions such as, for example, tropical cyclone wind field computations with `climada.trop_cyclone`". Or, alternatively, mention a single example of a run that you did: "For example, applying the procedure to XX years of data in the YY region may take up to ZZ hours on a [modern laptop computer|high performance cluster]".

peanutfun replied:
I added hints about the compute time. Computing one flood footprint for a larger country takes "few minutes", usually including the data download.

tovogt commented:
I didn't mean to suggest that you should have a single wrapper function that covers an excessive amount of use cases ("all that functionality"), but instead have a wrapper function that covers at least the type of use case that is described in the tutorial notebook. That seems really quite simple to me, and will not take a lot of exploring and testing, compared to all the work that you have already put into this.

I mean, currently, there is no unit test for the `compute_hazard_series` function, right? And that function needs a yml file as an input. Just imagine you want to implement a unit test for `compute_hazard_series` for a bunch of "typical" yml files that you generate on the fly. That's basically what that wrapper function would do.

From my point of view, there are two possible interpretations of the current state:

1. The use case described in the tutorial notebook is very common. Then there should definitely be a wrapper function, because it's really very simple: all the modifications to the yml file mentioned in the tutorial are extremely trivial.
2. The use case described in the tutorial notebook is very artificial because you needed to break everything down to a point where you can actually execute the stuff in a Jupyter instance on your computer. In that case, we really don't need a wrapper function, or at least not a simple one. But then, I'm missing a section in the tutorial notebook that explains what **would** be a typical application, and at which point we would typically need to make non-trivial modifications to the yml-file.

In both cases, there would still be work to do here. I don't think that this PR is ready to be merged as is (or at least I still need to be convinced).

peanutfun commented:
@emanuel-schmid The docs for my module [are empty](https://climada-petals--64.org.readthedocs.build/en/64/climada_petals/climada_petals.hazard.rf_glofas.html) because the doc environment does not contain the new dependencies (`ruamel.yaml` and `xesmf`), and hence the documentation parser fails. How to proceed here? Could you add them?

emanuel-schmid commented:
> @emanuel-schmid The docs for my module [are empty](https://climada-petals--64.org.readthedocs.build/en/64/climada_petals/climada_petals.hazard.rf_glofas.html) because the doc environment does not contain the new dependencies (`ruamel.yaml`, `xesmf` and `cdsapi`), and hence the documentation parser fails. How to proceed here? Could you add them?

sure. [done](https://github.com/CLIMADA-project/climada_petals/commit/0d830cc3706e3c5f509a175264a13d7948ea31f6). however - it's somewhat unsettling that the build doesn't fail and we'll never know unless we actually read the docs. 😟

peanutfun commented:
@emanuel-schmid Thank you! 🙌 

Indeed. Unfortunately, the only option is to turn *every* Sphinx warning into an error, including issues with links, heading levels and so on... This would be way to much work to resolve. As far as I can see, there is no feature that only turns module import errors into Sphinx build errors. I raised a feature request over at Sphinx: https://github.com/sphinx-doc/sphinx/issues/11764

peanutfun commented:
@tovogt @aleeciu I more-or-less rewrote the entire module. I removed the dependency on "dantro" and instead write the `RiverFloodInundation` class which handles the data transformation pipeline. This should make the actual computation steps much clearer. It still uses the transformations defined in `transform_ops`, which makes its methods a bit repetitive, but that was easier than rewriting everything. I updated the tutorial and documentation, and most of the code should be tested now.

I would be happy about further reviews, and I would like to merge this in December at the latest. Then it would align nicely with the submission of my paper.

peanutfun commented on setup.py:
```diff
@@ -46,7 +46,11 @@
+        'dantro',
+        'ruamel.yaml',
         'scikit-image',
+        'xesmf',
```
I fixed the conda specs.

tovogt reviewed (commented):
I went through the unit tests and tried to address all warnings that popped up.

Then, I spent some time going through the docs and tutorial, and scanning the code, and I had the following thoughts:

1. It's really unfortunate that data needs to be written uncompressed first due to performance issues. Actually, compressing with `zlib=True` is usually really very, very fast, but it can be become ridiculously slow when using dask, see https://github.com/pydata/xarray/issues/2912. However, in the past I was always somehow able to deal with it by calling `compute()` at the right time. Unfortunately, I don't have time to dig deeper here, but I feel like this is something where improvement should be possible.
2. In the tutorial notebook, there are some occurrences of `py:func` which are not translated properly by Sphinx. Maybe this feature is not supported or the syntax should be different somehow?

Finally, a lot of function docstrings are incomplete. Some function arguments are not mentioned in the `Parameters` section, or the `Parameters` section is missing.

peanutfun commented on climada_petals/hazard/rf_glofas/test/test_rf_glofas.py:
```diff
@@ -0,0 +1,636 @@
+        # Assert calls
+        save_file(ds, outpath, encoding, **encoding_defaults)
+        ds.to_netcdf.assert_called_once_with(
+            outpath.with_suffix(".nc"),
+            encoding=dict(
+                foo=dict(dtype="float32", zlib=True, complevel=4, other_setting=False),
+                bar=dict(
+                    dtype="float64",
+                    zlib=True,
+                    complevel=4,
+                    other_setting=False,
+                    some_setting=True,
+                ),
+            ),
+            engine="netcdf4",
+        )
+        ds.to_netcdf.reset_mock()
+
+        # Any suffix will be forwarded
+        outpath = outpath.with_suffix(".suffix")
+        defaults = dict(dtype="float32", zlib=False, complevel=4)
+        save_file(ds, outpath)
+        ds.to_netcdf.assert_called_once_with(
+            outpath, encoding=dict(foo=defaults, bar=defaults), engine="netcdf4"
+        )
+
+        # KeyError for data_vars that do not exist
+        with self.assertRaises(KeyError) as cm:
+            save_file(ds, outpath, dict(baz=dict(stuff=True)))
+        self.assertIn("baz", str(cm.exception))
+        ds.to_netcdf.reset_mock()
+
+        # DataArray must be promoted
+        da = MagicMock(xr.DataArray)
+        da.to_dataset.return_value = ds
+        save_file(da, outpath)
+        da.to_dataset.assert_called_once_with()
+        ds.to_netcdf.assert_called_once_with(
+            outpath, encoding=dict(foo=defaults, bar=defaults), engine="netcdf4"
+        )
```
No need as far as I see. I only use the temporary directory to have a temporary directory *name*. The dataset is mocked, and no file is ever written. We could alternatively just use a custom file name.

tovogt replied:
Either use a custom file name, or use the `TemporaryDirectory` properly. Otherwise, we get nasty warnings.

tovogt replied:
Alternatively, use `tempfile.mkdtemp` as suggested below in https://github.com/CLIMADA-project/climada_petals/pull/64#discussion_r1407812035

peanutfun commented on climada_petals/hazard/rf_glofas/test/test_river_flood_computation.py:
```diff
@@ -0,0 +1,381 @@
+        self.arr_2 = xr.DataArray(data=[1], coords={"dim": [1]})
+        self.filename = Path(self.tempdir.name) / "file.nc"
+        self.arr_2.to_netcdf(self.filename)
+
```
Note the [docs](https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory):

> On completion of the context or destruction of the temporary directory object, the newly created temporary directory and all its contents are removed from the filesystem. [...]
> The directory can be explicitly cleaned up by calling the cleanup() method.
> [...] errors will be raised in whatever context cleanup occurs (the cleanup() call, exiting the context manager, when the object is garbage-collected or during interpreter shutdown).

As far as I understand, the directory is cleaned up when the `TemporaryDirectory` object is destroyed or garbage-collected, which should happen the next time `setUp` is called or the entire test object is deleted. Explicitly calling cleanup should not be required. (But it might be good practice nonetheless?)

tovogt replied:
Whatever you do, get rid of the warnings :D

peanutfun replied:
@tovogt What warnings are you referring to? On Python 3.9, running pytest on the module I only get the `RuntimeWarning: divide by zero encountered in divide` you mentioned elsewhere. All other warnings I get are not related to my module, as far as I can see.

tovogt replied:
```shell
$ python --version
Python 3.9.16
$ python -m unittest climada_petals.hazard.rf_glofas.test.test_rf_glofas.TestTransformOps.test_save_file
/home/tovogt/.local/share/miniconda3/envs/climada_env_3.9/lib/python3.9/site-packages/dask/dataframe/_pyarrow_compat.py:17: FutureWarning: Minimal version of pyarrow will soon be increased to 14.0.1. You are using 11.0.0. Please consider upgrading.
  warnings.warn(
/home/tovogt/.local/share/miniconda3/envs/climada_env_3.9/lib/python3.9/tempfile.py:821: ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/tmp/tmpxkfriv_i'>
  _warnings.warn(warn_message, ResourceWarning)
.
----------------------------------------------------------------------
Ran 1 test in 0.003s

OK
```

tovogt replied:
My interpretation is that the whole point of `TemporaryDirectory` are the automatic cleanup features. If you don't want to make use of it, there is tempfile.mkdtemp which does the same without the automatic cleanup.

peanutfun replied:
See https://github.com/CLIMADA-project/climada_petals/pull/64#discussion_r1418964664

peanutfun commented on climada_petals/hazard/river_flood.py:
```diff
@@ -40,6 +41,8 @@
 from climada.hazard.base import Hazard
 from climada.hazard.centroids import Centroids
 
+# from .rf_glofas import GloFASRiverFlood
+
```
Yes, leftover from previous implemetation

peanutfun commented on climada_petals/hazard/river_flood.py:
```diff
@@ -40,6 +41,8 @@
 from climada.hazard.base import Hazard
 from climada.hazard.centroids import Centroids
 
+# from .rf_glofas import GloFASRiverFlood
+
```
```suggestion
```

peanutfun replied:
```suggestion
```

peanutfun commented on climada_petals/hazard/river_flood.py:
```diff
@@ -21,6 +21,7 @@
 
 __all__ = ['RiverFlood']
 
+from typing import Union
```
```suggestion
```

peanutfun commented on doc/conf.py:
```diff
@@ -247,6 +247,10 @@ def remove_module_docstring(app, what, name, obj, options, lines):
 
+# Mock these modules instead of importing them when creating the docs
+# TODO: Mocking modules should not be necessary when the docs environment is up-to-date
+autodoc_mock_imports = ["dantro", "cdsapi", "xesmf"]
```
Yes, and the doc environment is actually up-to-date now. So this can be removed.

peanutfun commented on doc/conf.py:
```diff
@@ -247,6 +247,10 @@ def remove_module_docstring(app, what, name, obj, options, lines):
 
 autodoc_member_order = "bysource"
 
+# Mock these modules instead of importing them when creating the docs
+# TODO: Mocking modules should not be necessary when the docs environment is up-to-date
+autodoc_mock_imports = ["dantro", "cdsapi", "xesmf"]
+
```
```suggestion
```

peanutfun commented on climada_petals/hazard/rf_glofas/transform_ops.py:
```diff
@@ -0,0 +1,809 @@
+    np.ndarray
+        The return period(s) for the input parameters
+    """
+    return np.minimum(1.0 / (1.0 - gumbel_r.cdf(sample, loc=loc, scale=scale)), max_rp)
```
Very good point! But wouldn't it be simpler (and safe because of floating point arithmetics) to *always* add the spacing?
```suggestion
    cdf = gumbel_r.cdf(sample, loc=loc, scale=scale)
    return np.fmin(1.0 / (1.0 - cdf + np.spacing(1.0)), max_rp)
```

tovogt replied:
In cases where `cdf == 1.0`, my version will return `np.inf`, and your version will return a very large float. In my opinion, the case `cdf == 1.0` is in a different category of what can happen, so I would also want the code to be able to produce `np.inf` as a return value in this case. If users don't like `np.inf`, they are free to set `max_rp` to anything that they prefer. But if users want to have `np.inf`, e.g. for doing sanity checks, they should be able to get this value by setting `max_rp` to `np.inf` (which happens to be the default, actually).

peanutfun commented on climada_petals/hazard/rf_glofas/test/test_rf_glofas.py:
```diff
@@ -0,0 +1,636 @@
+        # Test special sample values
+        samples = gev.copy()
+        samples[0, 0] = 0
+        samples[0, 1] = np.nan
+        samples[1, 0] = np.inf
```
Good catch with the data types! I made sure that all value arrays are floats in 719fa14757964f09f8b90500fa215939bba2f06c. For coordinates, this distinction should not be important

tovogt replied:
I assumed that `samples` contains the number of samples, so it should be an integer, right?

peanutfun replied:
Absolutely. You seem to know my tests better that I do 😅  I think I wanted to go for the `isfinite` check inside the method. If `samples` are integers, they should always be finite. So this does not make sense either way. What was catched instead was `samples < 1`. Will try to handle this correctly in an updated version of the test.

peanutfun commented on climada_petals/hazard/rf_glofas/test/test_rf_glofas.py:
```diff
@@ -0,0 +1,636 @@
+    def test_return_period_resample(self):
+        """Test 'return_period_resample' operation"""
+        # Make more than 2 dims to check handling of 'core_dims'
+        x = np.arange(10)
+        y = np.arange(20, 10, -1)
```
I casted `values` to float and left coordinates as ints. This should be fine.

peanutfun commented on climada_petals/hazard/rf_glofas/river_flood_computation.py:
```diff
@@ -0,0 +1,658 @@
+        cache_dir = Path(cache_dir)
+        cache_dir.mkdir(parents=True, exist_ok=True)
+
+        # Create temporary directory for cache
+        self._tempdir = TemporaryDirectory(
+            dir=cache_dir, prefix=datetime.today().strftime("%y%m%d-%H%M%S-")
+        )
+
+        # Define cache paths
+        self.cache_paths = RiverFloodCachePaths.from_dir(Path(self._tempdir.name))
+
+    def clear_cache(self):
+        """Clear the cache of this instance
+
+        This will delete the temporary cache directory and create a new one by calling
+        :py:meth:`_create_tempdir`.
+        """
+        cache_dir = Path(self._tempdir.name).parent
+        self._tempdir.cleanup()
+        self._create_tempdir(cache_dir=cache_dir)
```
I tried to find an elegant solution to this. I personally favor `TemporaryDirectory` over `mkdtemp` because it actually *tries* to clean up implicitly. But I see that I should not rely on it exclusively. So here's my proposal:

* Use `TemporaryDirectory`
* Call `cleanup` when I intend to clean up the directory before re-assigning the `_tempdir`.
* Add a `__del__` method which also calls `cleanup` to avoid implicit cleanup, although this [seems discouraged](https://stackoverflow.com/a/2452895/22190523).

I'll also add the appropriate calls to `cleanup` to the tests

tovogt replied:
That sounds a bit like you are reinventing the wheel. Maybe just copy what `TemporaryDirectory` does internally and use `weakref.finalize`: https://github.com/JakobDev/cpython/blob/92ac7ba794271f9437993b859799dc6f4355b1f8/Lib/tempfile.py#L869

peanutfun replied:
Don't you think re-implementing internals of `TemporaryDirectory` would be reinventing the wheel? I think the code remains much more understandable when we stick to `TemporaryDirectory`. Please have a look at the latest version, I think it should be both safe and understandable.

peanutfun replied:
@tovogt How to continue here? I think the current implementation using `__del__` is simple and clean.

tovogt replied:
I still think, it's not unsatisfactory that using `__del__` cannot guarantee to suppress the warnings by `TemporaryDirectory`. But admittedly, using weakref feels clumsy since you have to define a callback and keep references of it around etc. So, I agree that `__del__` is simple, but I don't think it's clean. Still, I agree to keep it like it is.

peanutfun commented on climada_petals/hazard/rf_glofas/test/test_rf_glofas.py:
```diff
@@ -0,0 +1,636 @@
+import geopandas as gpd
+from shapely.geometry import Polygon
+
+from climada_petals.hazard.rf_glofas.transform_ops import (
```
Correct. I renamed the file and added a new file `test_rf_glofas.py` with test for that particular module.

peanutfun commented on climada_petals/hazard/rf_glofas/rf_glofas.py:
```diff
@@ -0,0 +1,166 @@
+        client.close()
+
+
+def hazard_series_from_dataset(
+    data: xr.Dataset, intensity: str, event_dim: str
+) -> Union[pd.Series, Hazard]:
```
I added tests for `hazard_series_from_dataset` based on your suggestion in aa0cdd2253f5f0d6e4a920d5cfc7bf96f2e26419. For me, your tests actually passed!

tovogt replied:
Your tests only ever test the case where the `event_dim` variable is a datetime dimension. But the example given in your docstring uses `"number"` as the `event_dim` and in that case, I would not expect this dimension to be datetime-like.

Accordingly, this is what I (still) get when running my suggested test:
```Python
======================================================================
ERROR: test_hazard_series_from_dataset (climada_petals.hazard.rf_glofas.test.test_rf_glofas.TestHazardSeriesFromDataset)
Test the hazard_series_from_dataset function
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tovogt/code/climada_petals/climada_petals/hazard/rf_glofas/test/test_rf_glofas.py", line 118, in test_hazard_series_from_dataset
    haz_series = hazard_series_from_dataset(ds, "intensity", "number")
  File "/home/tovogt/code/climada_petals/climada_petals/hazard/rf_glofas/rf_glofas.py", line 160, in hazard_series_from_dataset
    hazards = [
  File "/home/tovogt/code/climada_petals/climada_petals/hazard/rf_glofas/rf_glofas.py", line 161, in <listcomp>
    create_hazard(data.sel(dict(zip(iter_dims, idx))))
  File "/home/tovogt/code/climada_petals/climada_petals/hazard/rf_glofas/rf_glofas.py", line 144, in create_hazard
    return Hazard.from_xarray_raster(
  File "/home/tovogt/code/climada_python/climada/hazard/base.py", line 843, in from_xarray_raster
    data[coords["event"]].dt.strftime("%Y-%m-%d").values.flatten().tolist(),
  File "/home/tovogt/.local/share/miniconda3/envs/climada_env_3.9/lib/python3.9/site-packages/xarray/core/utils.py", line 1108, in __get__
    return self._accessor(obj)  # type: ignore  # assume it is a valid accessor!
  File "/home/tovogt/.local/share/miniconda3/envs/climada_env_3.9/lib/python3.9/typing.py", line 687, in __call__
    result = self.__origin__(*args, **kwargs)
  File "/home/tovogt/.local/share/miniconda3/envs/climada_env_3.9/lib/python3.9/site-packages/xarray/core/accessor_dt.py", line 623, in __new__
    raise TypeError(
TypeError: '.dt' accessor only available for DataArray with datetime64 timedelta64 dtype or for arrays containing cftime datetime objects.

----------------------------------------------------------------------
```

peanutfun replied:
Indeed, I get the same error. It seems like https://github.com/CLIMADA-project/climada_python/pull/795 broke my code because it now requires either explicit event names or that the "event" dimension can be interpreted as DateTime, which was not a requirement before.

peanutfun replied:
https://github.com/CLIMADA-project/climada_python/pull/837 should have resolved this. I'll add an appropriate test case.

peanutfun commented:
@tovogt Thanks again for the thorough review! To comment on your overall thoughts:

> It's really unfortunate that data needs to be written uncompressed first due to performance issues.

Yes, but this is what I came up with after months of using the module. I don't think I can do much better without further help. In my experience, using `zlib` is indeed horribly slow with dask, and it also does not properly support multi-process writing to a single file. You suggest calling `compute` first and then saving the data. However, this requires the entire data to fit into memory in the first place. Depending on the country and time frame you want to compute flood footprints for, this is not feasible given the usual memory space of a personal computer.

However, my module actually gives users the option to optimize this themselves. Setting `store_intermediates` to `False` and executing each step of `RiverFloodComputation.compute` themselves, they are free to call `compute` and store data as they see fit. I tried to find default settings that work no-matter-what. The main restriction now is drive space. You might end up writing hundreds of GB, but it actually is quite performant.

> In the tutorial notebook, there are some occurrences of py:func which are not translated properly by Sphinx.

The Myst parser should support these types of references, see https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#reference-roles

Now that I look at the `doc/conf.py`, we might actually still use the old nbsphinx parser for reading the notebooks. I'll try to fix that.

tovogt commented:
> Yes, but this is what I came up with after months of using the module. I don't think I can do much better without further help. In my experience, using `zlib` is indeed horribly slow with dask, and it also does not properly support multi-process writing to a single file. You suggest calling `compute` first and then saving the data. However, this requires the entire data to fit into memory in the first place. Depending on the country and time frame you want to compute flood footprints for, this is not feasible given the usual memory space of a personal computer.

I think it's desirable to require each individual NetCDF file to contain at most as much data as could potentially fit into (a reasonable amount of) memory. Ideally, I would even propose to have at most 4 GB of (uncompressed) data per NetCDF file. With NetCDFs, it is very easy to just split up data into several files, and then load the dataset as a multi-file-dataset, e.g. using `xr.open_mfdataset`. If you adhere to that, you can very well call `compute` for each chunk of data that's supposed to end up in an individual file. It's much faster, the data is very easy to handle, and there are practically no disadvantages. Monolithic, long-running processes that produce monolithic chunks of data, are extremely inconvenient in almost every environment. For all projects I work with, I try to split up everything into a high number of slim processes with short run times that produce comparably small chunks of data each - and it's so much more convenient in almost any circumstance I can think of.

> However, my module actually gives users the option to optimize this themselves. Setting `store_intermediates` to `False` and executing each step of `RiverFloodComputation.compute` themselves, they are free to call `compute` and store data as they see fit. I tried to find default settings that work no-matter-what. The main restriction now is drive space. You might end up writing hundreds of GB, but it actually is quite performant (with a modern SSD)

As I said, this is not a merge-blocker from my side, and I'm happy to go with this solution.

tovogt commented on climada_petals/hazard/rf_glofas/test/test_transform_ops.py:
```diff
@@ -0,0 +1,662 @@
+        y = np.arange(20, 10, -1)
+        z = np.linspace(0, 5, 11)
+
+        # NOTE: np.multiply.outer does not flatten input
+        values = np.multiply.outer(np.outer(x, y), z).astype("float")
```
Easier to read, in my opinion, but no strong feelings:
```suggestion
        values = (x[:, None, None] * y[None, :, None] * z[None, None, :]).astype("float")
```

tovogt commented on climada_petals/hazard/rf_glofas/transform_ops.py:
```diff
@@ -0,0 +1,817 @@
+    # NOTE: 'rp_sampling' requires scalar 'loc' and 'scale' parameters, so we
+    #       define all but 'longitude' and 'latitude' dimensions as core dimensions
+    # core_dims = set(discharge.dims) - {"longitude", "latitude"}
+    ret = xr.apply_ufunc(
+        rp_sampling,
+        discharge,
+        gev_loc,
+        gev_scale,
+        gev_samples,
+        max_return_period,
+        input_core_dims=[list(core_dims), [], [], [], []],
+        output_core_dims=[["sample"] + list(core_dims)],
+        vectorize=True,
+        dask="parallelized",
+        output_dtypes=[np.float32],
+        dask_gufunc_kwargs=dict(
+            output_sizes=dict(sample=bootstrap_samples), allow_rechunk=True
+        ),
+    ).rename("Return Period")
+    ret = ret.assign_coords(sample=np.arange(bootstrap_samples))
+
+    # Transpose makes for a nicer result?
+    dims = list(ret.sizes.keys())
+    dims.remove("sample")
+    dims.append("sample")
+    ret = ret.transpose(*dims)
+
+    # Return result
+    return ret
```
```suggestion
    return (
        xr.apply_ufunc(
            rp_sampling,
            discharge,
            gev_loc,
            gev_scale,
            gev_samples,
            max_return_period,
            input_core_dims=[list(core_dims), [], [], [], []],
            output_core_dims=[["sample"] + list(core_dims)],
            vectorize=True,
            dask="parallelized",
            output_dtypes=[np.float32],
            dask_gufunc_kwargs=dict(
                output_sizes=dict(sample=bootstrap_samples), allow_rechunk=True
            ),
        )
        .rename("Return Period")
        .assign_coords(sample=np.arange(bootstrap_samples))
        .transpose(..., "sample")
    )
```

tovogt commented on climada_petals/hazard/rf_glofas/test/test_transform_ops.py:
```diff
@@ -0,0 +1,662 @@
+            return_period_resample(discharge, gev, gev, samples, bootstrap_samples)
+
+            # Test number of calls
+            print(samples.values)
```
Please remove print statement

tovogt commented:
You are using `xesmf` to regrid a raster with bilinear interpolation. Why don't you use `rasterio.warp.reproject` for that? It would avoid introducing a new dependency, it's really very powerful, and it's already used in several other places in CLIMADA.

peanutfun commented:
> Why don't you use rasterio.warp.reproject for that?

Simply because I am not familiar with it. I first used the xarray-internal interpolate, which is horribly slow and does not take geospatial information into account. So I switched to xesmf because it was recommended to me and is simple to use for xarray data structures. But it is also difficult to install, and poses an issue on Euler. So I would be very happy about a suggestion how to drop it and switch to another implementation, given that it is not much slower.

tovogt commented:
Okay, after a closer look, I think you won't be able to have anything similar to `nearest_s2d` extrapolation in `rasterio` or similarly basic packages.

peanutfun commented on climada_petals/hazard/rf_glofas/cds_glofas_downloader.py:
```diff
@@ -0,0 +1,286 @@
+
+Functions for downloading GloFAS river discharge data from the Copernicus Climate Data
+Store (CDS).
+"""
```
TODO: Update docs in this file

ThomasRoosli replied:
I updated the docstrings of the functions to now include the `use_case` parameter as well as all other missing parameters.

tovogt reviewed (approved):
Thanks @ThomasRoosli for the final cleanup. This is ready to be merged from my side.
